### PR TITLE
feat(notes): save chat answers with citation hover anchors (closes #660)

### LINKED ISSUE #660: Notes created via CLI (note create / --save-as-note) lack interactive [N] citation hover links
## Summary

Notes created via the CLI (`notebooklm note create` or `ask --save-as-note`) are plain text and do NOT have interactive `[N]` citation hover links. In contrast, when a user manually saves a conversation answer to a note via the NLM web UI's "Save to note" button, the `[N]` references become clickable with hover popups showing the exact source passage.

## Steps to Reproduce

1. `notebooklm ask --json -c <server_conv_id> "question"` → answer appears in web UI with working `[N]` hover links ✓
2. `notebooklm ask --json -c <server_conv_id> --save-as-note --note-title "Test" "question"` → note is created but `[N]` links are plain text, no hover ✗
3. Manually save the same answer via web UI "Save to note" → `[N]` hover links work ✓

## Analysis

The CLI's `--save-as-note` implementation calls `client.notes.create(notebook_id, title, result.answer)` which creates a note from raw answer text. The web UI's "Save to note" feature likely uses a different internal mechanism that associates the note content with source passage metadata (chunk_id, start_char, end_char from the chat response's `references` field).

The `AskResult.references` already contains the necessary data:
```json
{
  "source_id": "c1747023-...",
  "citation_number": 1,
  "cited_text": "relevant passage text...",
  "start_char": 25617,
  "end_char": 25716,
  "chunk_id": "a2a525fc-..."
}
```

## Proposed Enhancement

When `--save-as-note` is used, instead of `notes.create()` with plain text, use the same RPC/mechanism that the web UI uses to save a chat answer as a note — one that preserves the citation-to-source-passage associations. This would make `[N]` references in CLI-created notes interactive, matching the web UI behavior.

## Impact

This is important for academic research workflows where users need to verify cited source passages. Currently, users must:
1. Use CLI for automated Q&A and citation processing
2. Then manually go to the web UI to save each answer as a note (to get interactive citations)

If `--save-as-note` preserved `[N]` hover links, the entire workflow could be automated.

## Related

- #659 (CLI conversations invisible in web UI — workaround: use `-c <server_conv_id>`)

## Environment

- notebooklm-py version: 0.4.1
- Platform: WSL2 Ubuntu, Python 3.12

### PR DESCRIPTION
**Pull request by teng-lin**
## Summary

- Closes #660: `notebooklm ask --save-as-note` now produces notes with **interactive `[N]` hover-anchored citations**, matching the NotebookLM web UI's "Save to note" button. Previously the saved note was plain text with `[N]` rendered as literal characters.
- Adds `NotesAPI.create_from_chat(notebook_id, ask_result, *, title=None)` as the public entry point; under the hood it uses a single-shot **`CREATE_NOTE` (saved-from-chat variant)** with a 7-element params array and mode flag `[2]` — wire format reverse-engineered from a captured web-UI curl request and pinned by a **byte-exact golden test**.
- CLI `--save-as-note` automatically routes through the citation-rich path when the answer has references; falls back to plain-text `notes.create()` (with a yellow hint in the terminal) when it doesn't, so existing callers keep working.

## Why a new method instead of extending `notes.create()`

`notes.create()` has a different wire shape (5-element blank-note `CREATE_NOTE` + follow-up `UPDATE_NOTE`) and stable callers. The citation-rich save is a different RPC payload (mode flag `[2]` vs `[1]`, no `UPDATE_NOTE`), needs an `AskResult` (not free-form text), and is opt-in. Keeping it as a sibling method preserves `notes.create()`'s contract and lets the CLI pick the right path based on whether citations exist.

## Key design decisions / open questions

| Topic | Decision | Notes |
|---|---|---|
| 4th UUID at wire slot `[3][0][5][0][0]` (`06110a19-...` in capture) | Added optional `ChatReference.passage_id`; encoder uses it when set, falls back to `chunk_id` placeholder | This UUID does NOT appear in the streaming chat response we currently parse. The forward-compat field lets us populate it cleanly if a future capture reveals its source. Empirically the server accepts `chunk_id` here (manual test required for hover verification — see Test plan). |
| Boolean encoding (`0` vs `false`) | Integer `0` (NOT Python `False`) | `json.dumps(False) → "false"` but the captured wire payload uses `"0"`. Guarded by `test_golden_serializes_booleans_as_zero_not_false`. The server normalizes both; the request channel is strict. |
| Multi-citation segmentation | Cumulative-span heuristic: each `[N]` anchors `clean_text[0..position]` | Single-citation capture matches exactly. Multi-citation unverified — captured payload only had one citation. Documented in `docs/rpc-reference.md`. |
| Server smart-title | We send the requested title; surface server-returned title in `Note.title` | Captured response showed the server replaced `"New Saved Note"` with a smart French title. Documented in CLI/API docs. |
| `history --save` | Untouched (deferred) | Multi-turn citation handling isn't in the capture; stays on plain-text path. |

## Verification done

- `uv run ruff format . && uv run ruff check . && uv run mypy src/notebooklm --ignore-missing-imports && uv run pytest` — **4541 passed, 12 skipped (auth-gated E2E), 0 failures**.
- 18 encoder unit tests (golden + behavior), 5 high-level API tests, 5 CLI tests (3 existing + 2 new for citation-rich / fallback paths), 1 wire round-trip integration test (httpx-mock + captured response → asserts encoder output matches captured request byte-for-byte and `Note` parses correctly from captured response), 1 E2E stub gated by `@requires_auth`.

## Test plan

- [ ] CI (ruff/mypy/pytest) green on this PR.
- [ ] **Manual hover-anchor verification** (cannot be automated; requires live auth): run `notebooklm ask --save-as-note --note-title "hover-test" "<question_with_citations>"` against a real notebook, open the resulting note in https://notebooklm.google.com/, hover any `[N]` marker, **confirm a popup appears with the cited passage text**. If hover renders plain text without a popup, the 4th-UUID placeholder isn't accepted by the server and we need a fresh capture — file follow-up against #660.
- [ ] Verify `ask --save-as-note` on an answer with NO citations falls back to plain-text save with the yellow `[dim]No citations in answer...[/dim]` hint.

## Deferred polish (from code-simplifier review)

Tracked here for future cleanup; none block merge:
1. `_strip_citation_markers` clean-offset accumulation could use `len("".join(parts))` — cosmetic.
2. `NotesAPI.create_from_chat` raises `ValueError` on empty refs, but the CLI pre-checks and never reaches it. Defensive at the library API; acceptable.
3. Duplicate AskResult-from-fixture reconstruction lives in 2 test files — could extract a helper, only 2 call sites today.
4. Title-default truncation uses magic number `50` — matches existing project patterns.

## Out of scope / follow-ups

- `history --save` multi-turn citation preservation (no capture available).
- Surfacing the 4th UUID via the chat parser if/when its source is identified.
- Multi-citation segmentation verification against a captured payload with `[1][2][3]` markers.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  - Save chat answers as citation-rich notes that preserve interactive hover-anchored citation links; server may generate a “smart” title that overrides a requested title and is shown in the success message.
  - CLI now routes saves: citation-rich answers use the citation-preserving path; answers without citations fall back to plain-text saves and show a dim “no citations” notice.

* **Documentation**
  - Expanded CLI, Python API, and RPC docs describing citation-rich note saving behavior and title/encoding nuances.

* **Tests**
  - Added unit, integration, and e2e tests covering citation-preserving saves and edge cases.

[Review Change Stack](https://app.coderabbit.ai/change-stack/teng-lin/notebooklm-py/pull/675?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

**Comment by teng-lin** [comment]
@claude review

File: src/notebooklm/_notes.py [hunk 1]
```diff
@@ -109,6 +109,59 @@ async def create(
...
+
+        Raises:
+            ValueError: If ``ask_result.references`` is empty.
+        """
+        if not ask_result.references:
+            raise ValueError(
+                "create_from_chat requires AskResult.references to be "
+                "non-empty; use notes.create() for plain-text notes."
+            )
+        resolved_title = title if title is not None else f"Chat: {ask_result.answer[:50]}"
```
[the remark this answers is not in the document]
Comment (teng-lin) [reply src/notebooklm/_notes.py]: Addressed in 63d94e3 — default title now does `f"Chat: {ask_result.answer[:50].strip().replace(chr(10), ' ')}"` (slice before strip/replace, per your reference). Same sanitization applied to the CLI default in cli/chat.py.

File: src/notebooklm/cli/chat.py [hunk 2]
```diff
@@ -223,7 +232,19 @@ async def _run():
                         return
                     try:
                         title = note_title or f"Chat: {question[:50]}"
```
[the remark this answers is not in the document]
Comment (teng-lin) [reply src/notebooklm/cli/chat.py]: Addressed in 63d94e3 — applied the same slice-then-strip-then-replace pattern in cli/chat.py:234. Skipped the shared-helper extraction for now (would touch CLI helpers module outside this PR's scope); follow-up if other call sites grow.

File: tests/e2e/test_notes.py [hunk 3]
```diff
@@ -105,3 +105,57 @@ async def test_list_mind_maps(self, client, read_only_notebook_id):
...
+    These tests exercise the full ask → create_from_chat path against
+    the live NotebookLM service. The note's creation success is
+    automated; the per-citation hover-anchor rendering is a MANUAL
+    verification step — open the resulting note in the web UI, hover a
+    ``[N]`` marker, and confirm a passage popup appears. The test
+    prints instructions to the log to make this step actionable.
+    """
+
+    @pytest.mark.asyncio
+    async def test_create_from_chat_preserves_citations(self, client, temp_notebook):
```
[the remark this answers is not in the document]
Comment (teng-lin) [reply tests/e2e/test_notes.py:124]: Addressed in 63d94e3 — added @pytest.mark.e2e on TestCreateFromChat (alongside the existing @requires_auth). pytest was already imported.

File: tests/unit/test_save_chat_as_note_integration.py [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+    # Assert the request body carried the exact captured params.
+    request = httpx_mock.get_request()
+    body = request.content.decode()
+    form = parse_qs(body)
+    f_req = form["f.req"][0]
+    outer = json.loads(f_req)
+    rpc_id, inner_params_json, _null, _generic = outer[0][0]
+    assert rpc_id == RPCMethod.CREATE_NOTE.value
+    actual_params = json.loads(inner_params_json)
+    assert actual_params == request_params
```
[the remark this answers is not in the document]
Comment (teng-lin) [reply tests/unit/test_save_chat_as_note_integration.py:113]: Addressed in 63d94e3 — kept the existing `actual_params == request_params` deep-equal (good for failure-diff readability) and added a second assert comparing `json.dumps(..., separators=(',',':'))` strings. That catches the False==0 case you flagged.

**Comment by teng-lin** [comment]
@claude — review feedback addressed in 63d94e3. Per-finding response:

**Applied:**
- 🟡 `_TEXT_RENDER_FLAGS` mutable list → tuple at module level; call site copies via `list(_TEXT_RENDER_FLAGS)` when embedding. The golden test still passes (compact-JSON form is unchanged) and accidental mutation of an emitted params tree can no longer corrupt the module constant.
- 🟡 `passage_id or chunk_id` → `ref.passage_id if ref.passage_id is not None else ref.chunk_id`. Explicit None-check now distinguishes "intentionally empty string" from "unset".
- 🟡 Unit mocks vs wrapped response shape → `test_default_title_derives_from_answer` now mocks the wrapped form `[[note_id, ..., title, ...]]` matching the captured server response. Other tests in `TestCreateFromChat` keep the flat shape to retain defensive-branch coverage for the `elif isinstance(result[0], str)` fallback in `save_chat_answer_as_note`.

**Pushing back (rationale below):**
- 🟢 `_build_source_passage_descriptor` slot [6] can produce `[None]` — agreed in theory, but the only caller is `build_save_chat_as_note_params`, which guards via `seen_chunks: list[str]` (line ~358, only refs with truthy `chunk_id` enter the seen-list). Module-level callers don't exist today. Per the project's minimal-change / no-defensive-validation-for-impossible-states rule, adding a redundant guard inside the helper would be defensive-without-purpose. Happy to add it if a future caller bypasses the encoder's guard.
- 🟢 API title default uses answer-prefix, CLI uses question-prefix — this is intentional. The API caller has an `AskResult` (answer is the natural derivation), while the CLI caller has the original question and never reaches the API default. Both produce valid titles for their respective contexts. A docstring note could clarify but feels like rehashing what the function signatures already document.
- 🟢 `_strip_citation_markers("[1] starts here.")` leaves a leading space — yes, the regex `r" ?\[(\d+)\]"` consumes the optional leading space, not trailing. `test_marker_at_start` documents this as a known edge case. Production NotebookLM answers don't start with `[N]` (citations follow factual claims). If this turns into a real problem we can swap to `r" ?\[(\d+)\] ?"` but it'd risk eating spaces between adjacent words elsewhere.

Positives noted — thank you for calling out the golden test + fallback UX + `passage_id` forward-compat field. All three were deliberate design choices.

Also separately addressed via `polish(retrofit): #660` in 86172f8 (a triple-review pass after the original PR opened): Codex caught a critical source-vs-local span bug in `_build_source_passage_descriptor` that was masked by the happy-path fixture, and the boolean-serialization test was asserting on the fixture instead of the encoder output. Both fixed there.
